src/base/FeesManager.sol

```diff
@@ -0,0 +1,183 @@
+ *
+ *            (cumulatedFees - lastCumulatedFees) ⋅ shares
+ *    fees = ──────────────────────────────────────────────
+ *                                WAD
+ *
+ */
+abstract contract FeesManager is ReentrancyGuard {
+    /// @notice Cumulated fees for the Uniswap V4 pool `poolId` denominated in token0
+    mapping(PoolId poolId => uint256 cumulatedFees0) public getCumulatedFees0;
+
+    /// @notice Cumulated fees for the Uniswap V4 pool `poolId` denominated in token1
+    mapping(PoolId poolId => uint256 cumulatedFees1) public getCumulatedFees1;
+
+    /// @notice Last `collectFees` call of a `beneficiary` for a specific `poolId` denominated in token0
+    mapping(PoolId poolId => mapping(address beneficiary => uint256 lastCumulatedFees0)) public getLastCumulatedFees0;
+
+    /// @notice Last `collectFees` call of a `beneficiary` for a specific `poolId` denominated in token1
+    mapping(PoolId poolId => mapping(address beneficiary => uint256 lastCumulatedFees1)) public getLastCumulatedFees1;
+
+    /// @notice Shares entitled to a `beneficiary` for the associated Uniswap V4 `poolId`
+    mapping(PoolId poolId => mapping(address beneficiary => uint256 shares)) public getShares;
+
+    /// @notice Associates `poolId` types with their corresponding Uniswap V4 pool `poolKey` types
+    mapping(PoolId poolId => PoolKey poolKey) public getPoolKey;
+
+    /**
+     * @notice Collects fees from a locked Uniswap V4 pool, distributes to the caller if applicable
+     * @dev Collected fees are now held in this contract until they are claimed by their beneficiary
+     * @return fees0 Total fees collected in token0 since last collection
+     * @return fees1 Total fees collected in token1 since last collection
+     */
+    function collectFees(
+        PoolId poolId
+    ) external nonReentrant returns (uint128 fees0, uint128 fees1) {
+        BalanceDelta fees = _collectFees(poolId);
+        fees0 = uint128(fees.amount0());
+        fees1 = uint128(fees.amount1());
+
+        getCumulatedFees0[poolId] += fees0;
+        getCumulatedFees1[poolId] += fees1;
+
+        _releaseFees(poolId, msg.sender);
+    }
+
+    /**
+     * @notice Transfers the shares of the `msg.sender` to a new `newBeneficiary` address for a specified
+     * `poolId` Uniswap V4 pool, note that this function also collects the currently available fees
+     * @param poolId Pool id of the Uniswap V4 pool to target
+     * @param newBeneficiary Address of the new beneficiary
+     */
+    function updateBeneficiary(PoolId poolId, address newBeneficiary) external nonReentrant {
+        _releaseFees(poolId, msg.sender);
+        _releaseFees(poolId, newBeneficiary);
+
+        // No need to check if shares > WAD, since we already validated this in `_storeBeneficiaries`
+        getShares[poolId][newBeneficiary] += getShares[poolId][msg.sender];
+        getShares[poolId][msg.sender] = 0;
```

Comment:
In case `newBeneficiary == msg.sender`, the caller's shares will be deleted. Would be good to revert in this case to be safe